### PR TITLE
TST: best fun should be OptimizeResult.fun

### PR DESCRIPTION
**Pull request by andyfaff**
In #12111 the poster reports that the best function value (and best x) traced by the objective is greater than that of the `OptimizeResult`. In the example ('TNC' with `rosen` 5) there is a significant difference between the two. They should probably be the same.

This PR develops a test for this, and will be a place for fixes.

### CODE REVIEW

**Comment by mdhaber** [comment]
Hmm this looks like a good thing to have. Can't see the old CI logs; I'll close and open to re-run.

**Comment by andyfaff** [comment]
@mdhaber how does one xfail individual entries of a parameterised test?

**Comment by mdhaber** [comment]
I think this will help: https://docs.pytest.org/en/6.2.x/skipping.html#skip-xfail-with-parametrize

**Comment by mdhaber** [comment]
Looks like four methods are failing here: `cg`, `bfgs`, `l-bfgs-b`, and `slsqp` - but not by much. Which did you want to xfail?

**Comment by mdhaber** [comment]
I went ahead and xfailed the four methods that fail. I think just having this test is an improvement, so I'll go ahead and merge in a week if I don't hear otherwise. gh-15089 is still open, so the issue is still tracked.

File: scipy/optimize/tests/test_optimize.py [hunk 1]
```diff
@@ -1443,6 +1450,35 @@ def test_duplicate_evaluations(self, method):
                 raise RuntimeError(
                     "Duplicate evaluations made by {}".format(method))
 
+    def gh12111_cases():
```
Comment (mdhaber) [inline scipy/optimize/tests/test_optimize.py:1453]:
```suggestion
    def gh12111_cases(self):
```

File: scipy/optimize/tests/test_optimize.py [hunk 2]
```diff
@@ -1443,6 +1450,35 @@ def test_duplicate_evaluations(self, method):
...
+        xfail = pytest.mark.xfail(reason="not fixed yet")
+        for method in MINIMIZE_METHODS:
+            if method in {'cg', 'bfgs', 'l-bfgs-b', 'slsqp'}:
+                yield pytest.param(method, marks=xfail)
+            else:
+                yield method
+
+    @pytest.mark.parametrize('method', gh12111_cases())
```
Comment (mdhaber) [inline scipy/optimize/tests/test_optimize.py:1461]:
```suggestion
    @pytest.mark.parametrize('method', self.gh12111_cases())
```

**Comment by andyfaff** [comment]
I'm not really sure what to do with the minimisers that don't pass this test. How does one tell a minimiser to do extra cycles to get a smaller amount better when it's already thinks it's converged? If we report `best_x`/`best_fun` then things like the reported grad/hess may be slightly incorrect, and possibly (in the case of SLSQP) outside the bounds/constraints. Some people would want to have `best_fun`/`best_x`, some people not. In this test the differences are far out in the last decimal places, but there may be tests were the discrepancy is more important. 
Perhaps its a documentation/won't fix kind of problem?

**Comment by mdhaber** [comment]
Personally, I don't think it's a problem unless the algorithm guarantees that it converges to the best point it ever visited. I don't think they make that guarantee, so I don't think it's a problem that solvers fail this test. 

As I think you're suggesting, we could wrap `fun` and use the wrapper to keep track of the best point. Then, at the end when all the constraints are checked (gh-10112), we could also check `best_x` to see if it satisfies the constraints, too. We could document that the reported gradient/hessian corresponds with `x`, not `best_x`.

I don't really think we should do that, though. This is easy enough to do in user-land and probably rare enough to be considered important that I think it can be left to users.

I'll merge this in the meantime!

**Comment by mdhaber** [comment]
Yeah now that I see how platform-depended this is, I question whether we should even add the test. I wouldn't want this to be a whack-a-mole sort of thing where new failures keep coming up every so often and we just silence them. I'm happy to consider this a wontfix unless there is evidence that this is a bug in the implementation of the algorithm.

**Comment by mdhaber** [comment]
> I'm happy to consider this a wontfix unless there is evidence that this is a bug in the implementation of the algorithm.

What do you think about that, @andyfaff?

**Comment by andyfaff** [comment]
I think we should probably call this won't fix.

**Pull request closed without merging**